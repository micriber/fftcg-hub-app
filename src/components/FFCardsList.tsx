import React from 'react';
import FFCardSimple from './FFCardSimple';
import {FlatList, StyleSheet, Text, View} from 'react-native';
import {Card} from '../services/api/card';

type Props = {
  isListView: boolean;
  cards?: Card[];
  onCardPress?: (card: Card) => void;
  onEndReached?: () => void;
  onRefresh?: () => void;
  displayOwnPin?: boolean;
  refreshing?: boolean;
};

const FFCardsList = ({
  isListView,
  cards = [],
  displayOwnPin = false,
  refreshing = false,
  onCardPress = () => {},
  onRefresh = () => {},
  onEndReached = () => {},
}: Props) => {
  return (
    <View
      style={[
        isListView || cards.length === 0
          ? styles.listContainer
          : styles.gridContainer,
      ]}>
      <FlatList
        initialNumToRender={2}
        numColumns={isListView ? 1 : 2}
        key={isListView ? 0 : 1}
        data={cards}
        keyExtractor={(item) => item.code}
        renderItem={({item}) => (
          <FFCardSimple
            card={item}
            isListView={isListView}
            displayOwnPin={displayOwnPin}
            onPress={onCardPress}
          />
        )}
        onEndReached={onEndReached}
        onEndReachedThreshold={1}
        refreshing={refreshing}
        onRefresh={onRefresh}
        ListEmptyComponent={
          !refreshing ? (
            <Text style={styles.emptyMessage}>No result</Text>
          ) : null
        }
      />
    </View>
  );
};

const styles = StyleSheet.create({
  gridContainer: {
    marginLeft: 20,
    flexDirection: 'row',
  },
  listContainer: {
    flexDirection: 'column',
    justifyContent: 'flex-start',
    width: '100%',
    flexGrow: 1,
  },
  emptyMessage: {
    textAlign: 'center',
    marginTop: '50%',
    width: '100%',
  },
});

export default FFCardsList;
