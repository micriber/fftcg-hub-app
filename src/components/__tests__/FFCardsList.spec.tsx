/**
 * @format
 */

import 'react-native';
import React from 'react';
import FFCardsList from '../FFCardsList';
import {render} from '@testing-library/react-native';
import {Text} from 'react-native';

const cards = [
  {
    code: '1-001H',
    name: 'Boulou',
    text: '《火》《2》《ダル》',
    type: '風',
    element: 'blue',
    id: '1234-1234-1234-1234',
    rarity: 'H',
    cost: '1000',
    power: '',
    job: 'Soutien',
    category1: '',
    category2: '',
    multicard: '',
    exBurst: '',
    userCard: [],
  },
  {
    code: '1-002H',
    name: 'Boulou',
    text: '《火》《3》《ダル》',
    type: '風',
    element: 'blue',
    id: '1234-1234-1234-1234',
    rarity: 'H',
    cost: '1000',
    power: '',
    job: 'Soutien',
    category1: '',
    category2: '',
    multicard: '',
    exBurst: '',
    userCard: [],
  },
];

// More about testing in this repository:
// https://github.com/bamlab/react-native-testing/blob/master/src/pages/Home/__tests__/Home.test.tsx
describe('[Component] FFCardsList', () => {
  it('renders correctly in grid view', async () => {
    const root = render(<FFCardsList cards={cards} />);

    expect(root.toJSON()).toMatchSnapshot();
  });

  it('renders correctly in list view', async () => {
    const root = render(<FFCardsList cards={cards} />);

    expect(root.toJSON()).toMatchSnapshot();
  });
});
