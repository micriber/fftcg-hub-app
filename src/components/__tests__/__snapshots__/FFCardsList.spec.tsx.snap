// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Component] FFCardsList renders correctly in grid view 1`] = `
<View
  style={
    Array [
      Object {
        "flexDirection": "row",
        "marginLeft": 20,
      },
    ]
  }
>
  <RCTScrollView
    ListEmptyComponent={
      <Text
        style={
          Object {
            "marginTop": "50%",
            "textAlign": "center",
            "width": "100%",
          }
        }
      >
        No result
      </Text>
    }
    data={
      Array [
        Object {
          "category1": "",
          "category2": "",
          "code": "1-001H",
          "cost": "1000",
          "element": "blue",
          "exBurst": "",
          "id": "1234-1234-1234-1234",
          "job": "Soutien",
          "multicard": "",
          "name": "Boulou",
          "power": "",
          "rarity": "H",
          "text": "《火》《2》《ダル》",
          "type": "風",
          "userCard": Array [],
        },
        Object {
          "category1": "",
          "category2": "",
          "code": "1-002H",
          "cost": "1000",
          "element": "blue",
          "exBurst": "",
          "id": "1234-1234-1234-1234",
          "job": "Soutien",
          "multicard": "",
          "name": "Boulou",
          "power": "",
          "rarity": "H",
          "text": "《火》《3》《ダル》",
          "type": "風",
          "userCard": Array [],
        },
      ]
    }
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={2}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    onContentSizeChange={[Function]}
    onEndReached={[Function]}
    onEndReachedThreshold={1}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onRefresh={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    refreshControl={
      <RefreshControlMock
        onRefresh={[Function]}
        refreshing={false}
      />
    }
    refreshing={false}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <RCTRefreshControl />
    <View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Object {
              "flexDirection": "row",
            }
          }
        >
          <View
            style={
              Array [
                Object {
                  "marginLeft": 15,
                  "marginTop": 3,
                },
                undefined,
              ]
            }
            testID="SimpleFFCard"
          >
            <View
              accessible={true}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "opacity": 1,
                }
              }
            >
              <View
                style={
                  Array [
                    Object {
                      "overflow": "hidden",
                    },
                    Array [
                      Object {
                        "height": 444.6666666666667,
                        "width": 300,
                      },
                      undefined,
                    ],
                  ]
                }
              >
                <FastImageView
                  onFastImageError={[Function]}
                  resizeMode="contain"
                  source={
                    Object {
                      "uri": "https://fftcg.cdn.sewest.net/images/cards/full/1-001H_fr.jpg",
                    }
                  }
                  style={
                    Object {
                      "bottom": 0,
                      "left": 0,
                      "position": "absolute",
                      "right": 0,
                      "top": 0,
                    }
                  }
                  testID="Image-1-001H"
                />
              </View>
            </View>
          </View>
          <View
            style={
              Array [
                Object {
                  "marginLeft": 15,
                  "marginTop": 3,
                },
                undefined,
              ]
            }
            testID="SimpleFFCard"
          >
            <View
              accessible={true}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "opacity": 1,
                }
              }
            >
              <View
                style={
                  Array [
                    Object {
                      "overflow": "hidden",
                    },
                    Array [
                      Object {
                        "height": 444.6666666666667,
                        "width": 300,
                      },
                      undefined,
                    ],
                  ]
                }
              >
                <FastImageView
                  onFastImageError={[Function]}
                  resizeMode="contain"
                  source={
                    Object {
                      "uri": "https://fftcg.cdn.sewest.net/images/cards/full/1-002H_fr.jpg",
                    }
                  }
                  style={
                    Object {
                      "bottom": 0,
                      "left": 0,
                      "position": "absolute",
                      "right": 0,
                      "top": 0,
                    }
                  }
                  testID="Image-1-002H"
                />
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;

exports[`[Component] FFCardsList renders correctly in list view 1`] = `
<View
  style={
    Array [
      Object {
        "flexDirection": "column",
        "flexGrow": 1,
        "justifyContent": "flex-start",
        "width": "100%",
      },
    ]
  }
>
  <RCTScrollView
    ListEmptyComponent={
      <Text
        style={
          Object {
            "marginTop": "50%",
            "textAlign": "center",
            "width": "100%",
          }
        }
      >
        No result
      </Text>
    }
    data={
      Array [
        Object {
          "category1": "",
          "category2": "",
          "code": "1-001H",
          "cost": "1000",
          "element": "blue",
          "exBurst": "",
          "id": "1234-1234-1234-1234",
          "job": "Soutien",
          "multicard": "",
          "name": "Boulou",
          "power": "",
          "rarity": "H",
          "text": "《火》《2》《ダル》",
          "type": "風",
          "userCard": Array [],
        },
        Object {
          "category1": "",
          "category2": "",
          "code": "1-002H",
          "cost": "1000",
          "element": "blue",
          "exBurst": "",
          "id": "1234-1234-1234-1234",
          "job": "Soutien",
          "multicard": "",
          "name": "Boulou",
          "power": "",
          "rarity": "H",
          "text": "《火》《3》《ダル》",
          "type": "風",
          "userCard": Array [],
        },
      ]
    }
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={2}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    onContentSizeChange={[Function]}
    onEndReached={[Function]}
    onEndReachedThreshold={1}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onRefresh={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    refreshControl={
      <RefreshControlMock
        onRefresh={[Function]}
        refreshing={false}
      />
    }
    refreshing={false}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <RCTRefreshControl />
    <View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Array [
              Object {
                "flexDirection": "row",
                "height": 444.6666666666667,
              },
              undefined,
            ]
          }
          testID="SimpleFFCard"
        >
          <View
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
          >
            <View
              style={
                Array [
                  Object {
                    "overflow": "hidden",
                  },
                  Array [
                    Object {
                      "height": 444.6666666666667,
                      "width": 300,
                    },
                    undefined,
                  ],
                ]
              }
            >
              <FastImageView
                onFastImageError={[Function]}
                resizeMode="stretch"
                source={
                  Object {
                    "uri": "https://fftcg.cdn.sewest.net/images/cards/full/1-001H_fr.jpg",
                  }
                }
                style={
                  Object {
                    "bottom": 0,
                    "left": 0,
                    "position": "absolute",
                    "right": 0,
                    "top": 0,
                  }
                }
                testID="Image-1-001H"
              />
            </View>
          </View>
          <View
            style={
              Array [
                Object {
                  "borderColor": "#000",
                  "borderWidth": 1,
                  "height": 444.6666666666667,
                  "width": 450,
                },
              ]
            }
          >
            <Text>
              Code: 
              1-001H
            </Text>
            <Text>
              Nom: 
              Boulou
            </Text>
            <Text>
              Type: 
              風
            </Text>
            <Text>
              Element: 
              blue
            </Text>
            <Text />
            <Text>
              《
              <Image
                source="test-file-stub"
                style={
                  Array [
                    Object {
                      "height": 20,
                      "width": 20,
                    },
                    Object {},
                  ]
                }
              />
              》
              <Text
                style={
                  Array [
                    Object {
                      "borderColor": "#333",
                      "borderRadius": 80,
                      "borderWidth": 100,
                      "fontSize": 18,
                      "height": 20,
                      "width": 20,
                    },
                  ]
                }
              >
                ❷
              </Text>
              《
              <Image
                source="test-file-stub"
                style={
                  Array [
                    Object {
                      "height": 20,
                      "width": 20,
                    },
                    Object {},
                  ]
                }
              />
              》
            </Text>
          </View>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Array [
              Object {
                "flexDirection": "row",
                "height": 444.6666666666667,
              },
              undefined,
            ]
          }
          testID="SimpleFFCard"
        >
          <View
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
          >
            <View
              style={
                Array [
                  Object {
                    "overflow": "hidden",
                  },
                  Array [
                    Object {
                      "height": 444.6666666666667,
                      "width": 300,
                    },
                    undefined,
                  ],
                ]
              }
            >
              <FastImageView
                onFastImageError={[Function]}
                resizeMode="stretch"
                source={
                  Object {
                    "uri": "https://fftcg.cdn.sewest.net/images/cards/full/1-002H_fr.jpg",
                  }
                }
                style={
                  Object {
                    "bottom": 0,
                    "left": 0,
                    "position": "absolute",
                    "right": 0,
                    "top": 0,
                  }
                }
                testID="Image-1-002H"
              />
            </View>
          </View>
          <View
            style={
              Array [
                Object {
                  "borderColor": "#000",
                  "borderWidth": 1,
                  "height": 444.6666666666667,
                  "width": 450,
                },
              ]
            }
          >
            <Text>
              Code: 
              1-002H
            </Text>
            <Text>
              Nom: 
              Boulou
            </Text>
            <Text>
              Type: 
              風
            </Text>
            <Text>
              Element: 
              blue
            </Text>
            <Text />
            <Text>
              《
              <Image
                source="test-file-stub"
                style={
                  Array [
                    Object {
                      "height": 20,
                      "width": 20,
                    },
                    Object {},
                  ]
                }
              />
              》
              <Text
                style={
                  Array [
                    Object {
                      "borderColor": "#333",
                      "borderRadius": 80,
                      "borderWidth": 100,
                      "fontSize": 18,
                      "height": 20,
                      "width": 20,
                    },
                  ]
                }
              >
                ❸
              </Text>
              《
              <Image
                source="test-file-stub"
                style={
                  Array [
                    Object {
                      "height": 20,
                      "width": 20,
                    },
                    Object {},
                  ]
                }
              />
              》
            </Text>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;

exports[`[Component] FFCardsList renders correctly with children 1`] = `
<View
  style={
    Array [
      Object {
        "flexDirection": "column",
        "flexGrow": 1,
        "justifyContent": "flex-start",
        "width": "100%",
      },
    ]
  }
>
  <RCTScrollView
    ListEmptyComponent={
      <Text
        style={
          Object {
            "marginTop": "50%",
            "textAlign": "center",
            "width": "100%",
          }
        }
      >
        No result
      </Text>
    }
    data={
      Array [
        Object {
          "category1": "",
          "category2": "",
          "code": "1-001H",
          "cost": "1000",
          "element": "blue",
          "exBurst": "",
          "id": "1234-1234-1234-1234",
          "job": "Soutien",
          "multicard": "",
          "name": "Boulou",
          "power": "",
          "rarity": "H",
          "text": "《火》《2》《ダル》",
          "type": "風",
          "userCard": Array [],
        },
        Object {
          "category1": "",
          "category2": "",
          "code": "1-002H",
          "cost": "1000",
          "element": "blue",
          "exBurst": "",
          "id": "1234-1234-1234-1234",
          "job": "Soutien",
          "multicard": "",
          "name": "Boulou",
          "power": "",
          "rarity": "H",
          "text": "《火》《3》《ダル》",
          "type": "風",
          "userCard": Array [],
        },
      ]
    }
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={2}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    onContentSizeChange={[Function]}
    onEndReached={[Function]}
    onEndReachedThreshold={1}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onRefresh={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    refreshControl={
      <RefreshControlMock
        onRefresh={[Function]}
        refreshing={false}
      />
    }
    refreshing={false}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <RCTRefreshControl />
    <View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Array [
              Object {
                "flexDirection": "row",
                "height": 444.6666666666667,
              },
              undefined,
            ]
          }
          testID="SimpleFFCard"
        >
          <View
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
          >
            <View
              style={
                Array [
                  Object {
                    "overflow": "hidden",
                  },
                  Array [
                    Object {
                      "height": 444.6666666666667,
                      "width": 300,
                    },
                    undefined,
                  ],
                ]
              }
            >
              <FastImageView
                onFastImageError={[Function]}
                resizeMode="stretch"
                source={
                  Object {
                    "uri": "https://fftcg.cdn.sewest.net/images/cards/full/1-001H_fr.jpg",
                  }
                }
                style={
                  Object {
                    "bottom": 0,
                    "left": 0,
                    "position": "absolute",
                    "right": 0,
                    "top": 0,
                  }
                }
                testID="Image-1-001H"
              />
            </View>
          </View>
          <View
            style={
              Array [
                Object {
                  "borderColor": "#000",
                  "borderWidth": 1,
                  "height": 444.6666666666667,
                  "width": 450,
                },
              ]
            }
          >
            <Text>
              Code: 
              1-001H
            </Text>
            <Text>
              Nom: 
              Boulou
            </Text>
            <Text>
              Type: 
              風
            </Text>
            <Text>
              Element: 
              blue
            </Text>
            <Text />
            <Text>
              《
              <Image
                source="test-file-stub"
                style={
                  Array [
                    Object {
                      "height": 20,
                      "width": 20,
                    },
                    Object {},
                  ]
                }
              />
              》
              <Text
                style={
                  Array [
                    Object {
                      "borderColor": "#333",
                      "borderRadius": 80,
                      "borderWidth": 100,
                      "fontSize": 18,
                      "height": 20,
                      "width": 20,
                    },
                  ]
                }
              >
                ❷
              </Text>
              《
              <Image
                source="test-file-stub"
                style={
                  Array [
                    Object {
                      "height": 20,
                      "width": 20,
                    },
                    Object {},
                  ]
                }
              />
              》
            </Text>
          </View>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Array [
              Object {
                "flexDirection": "row",
                "height": 444.6666666666667,
              },
              undefined,
            ]
          }
          testID="SimpleFFCard"
        >
          <View
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
          >
            <View
              style={
                Array [
                  Object {
                    "overflow": "hidden",
                  },
                  Array [
                    Object {
                      "height": 444.6666666666667,
                      "width": 300,
                    },
                    undefined,
                  ],
                ]
              }
            >
              <FastImageView
                onFastImageError={[Function]}
                resizeMode="stretch"
                source={
                  Object {
                    "uri": "https://fftcg.cdn.sewest.net/images/cards/full/1-002H_fr.jpg",
                  }
                }
                style={
                  Object {
                    "bottom": 0,
                    "left": 0,
                    "position": "absolute",
                    "right": 0,
                    "top": 0,
                  }
                }
                testID="Image-1-002H"
              />
            </View>
          </View>
          <View
            style={
              Array [
                Object {
                  "borderColor": "#000",
                  "borderWidth": 1,
                  "height": 444.6666666666667,
                  "width": 450,
                },
              ]
            }
          >
            <Text>
              Code: 
              1-002H
            </Text>
            <Text>
              Nom: 
              Boulou
            </Text>
            <Text>
              Type: 
              風
            </Text>
            <Text>
              Element: 
              blue
            </Text>
            <Text />
            <Text>
              《
              <Image
                source="test-file-stub"
                style={
                  Array [
                    Object {
                      "height": 20,
                      "width": 20,
                    },
                    Object {},
                  ]
                }
              />
              》
              <Text
                style={
                  Array [
                    Object {
                      "borderColor": "#333",
                      "borderRadius": 80,
                      "borderWidth": 100,
                      "fontSize": 18,
                      "height": 20,
                      "width": 20,
                    },
                  ]
                }
              >
                ❸
              </Text>
              《
              <Image
                source="test-file-stub"
                style={
                  Array [
                    Object {
                      "height": 20,
                      "width": 20,
                    },
                    Object {},
                  ]
                }
              />
              》
            </Text>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
